[PP] Optimize memory usage by releasing output memory earlier

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #153383

Considering `output_chunks` is only used for last stage, we should not keep the outputs of each stage in memory; this will allow memory to be freed earlier.

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

H-Huang committed:
[PP] Optimize memory usage by releasing output memory earlier

[ghstack-poisoned]

kwen2501 commented on torch/distributed/pipelining/stage.py:
```diff
@@ -719,7 +716,8 @@ def forward_one_chunk(
         output_tuple = _normalize_model_output_as_tuple(output)
 
         # Prepare for final output merge or reduction
-        self.output_chunks.append(output)
+        if self.is_last:
+            self.output_chunks.append(output)
```
nit: do you mind adding more comments here re why only the last stage needs it? Basically what's stated in the PR desc. Thanks!

kwen2501 reviewed (approved):
LGTM

H-Huang committed:
Update on "[PP] Optimize memory usage by releasing output memory earlier"


Considering `output_chunks` is only used for last stage, we should not keep the outputs of each stage in memory; this will allow memory to be freed earlier.

cc awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge

H-Huang commented:
@pytorchbot merge -i
